File: gui-cpp/src/pages/objecttablepage.cpp
/**
 * 页4: 对象字典表 — 实现。
 */
#include "objecttablepage.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QHeaderView>
#include <QBrush>
#include <QColor>

static const char *HEADERS[] = {
    "Index", "Name", "SubCount", "Category", "PDO Map", "SM Obj", "Risk", "CoeRead", "CoeWrite"
};
static const int NCOLS = 9;

ObjectTablePage::ObjectTablePage(QWidget *parent) : QWidget(parent) {
    auto *layout = new QVBoxLayout(this);

    // 过滤栏
    auto *filterRow = new QHBoxLayout;
    filterRow->addWidget(new QLabel("过滤:"));
    m_filterCombo = new QComboBox;
    m_filterCombo->addItems({"全部", "仅错误", "仅 callback", "仅 PDO", "仅 Fxxx", "已跳过"});
    connect(m_filterCombo, QOverload<int>::of(&QComboBox::currentIndexChanged),
            this, &ObjectTablePage::applyFilter);
    filterRow->addWidget(m_filterCombo);
    filterRow->addStretch();
    m_countLabel = new QLabel("0 objects");
    filterRow->addWidget(m_countLabel);
    layout->addLayout(filterRow);

    // 表格
    m_table = new QTableWidget;
    m_table->setColumnCount(NCOLS);
    QStringList headers;
    for (int i = 0; i < NCOLS; ++i) headers << HEADERS[i];
    m_table->setHorizontalHeaderLabels(headers);
    m_table->setEditTriggers(QTableWidget::NoEditTriggers);
    m_table->setSelectionBehavior(QTableWidget::SelectRows);
    m_table->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);
    layout->addWidget(m_table);
}

void ObjectTablePage::setObjects(const QList<QVariantMap> &objects) {
    m_objects = objects;
    applyFilter();
}

void ObjectTablePage::applyFilter() {
    int filter = m_filterCombo->currentIndex();
    QList<QVariantMap> filtered = m_objects;

    if (filter == 1) { // 仅错误
        filtered.erase(std::remove_if(filtered.begin(), filtered.end(),
            [](const QVariantMap &o) { return !o["risk"].toBool(); }), filtered.end());
    } else if (filter == 2) { // 仅 callback
        filtered.erase(std::remove_if(filtered.begin(), filtered.end(),
            [](const QVariantMap &o) { return o["CoeRead"].toString().isEmpty() && o["CoeWrite"].toString().isEmpty(); }), filtered.end());
    } else if (filter == 3) { // 仅 PDO
        filtered.erase(std::remove_if(filtered.begin(), filtered.end(),
            [](const QVariantMap &o) { return !o["pdo_map"].toBool(); }), filtered.end());
    } else if (filter == 4) { // 仅 Fxxx
        filtered.erase(std::remove_if(filtered.begin(), filtered.end(),
            [](const QVariantMap &o) { return o["index_int"].toInt() < 0xF000; }), filtered.end());
    } else if (filter == 5) { // 已跳过
        filtered.erase(std::remove_if(filtered.begin(), filtered.end(),
            [](const QVariantMap &o) { return !o["skipped"].toBool(); }), filtered.end());
    }

    m_table->setRowCount(filtered.size());
    for (int row = 0; row < filtered.size(); ++row) {
        const QVariantMap &obj = filtered[row];
        // 确定行颜色
        QColor color(255, 255, 255); // 白色
        if (obj["skipped"].toBool()) color = QColor(220, 220, 220);
        else if (obj["risk"].toBool()) color = QColor(255, 200, 200);
        else if (!obj["CoeRead"].toString().isEmpty() || !obj["CoeWrite"].toString().isEmpty())
            color = QColor(255, 255, 180);

        for (int col = 0; col < NCOLS; ++col) {
            auto *item = new QTableWidgetItem(obj[HEADERS[col]].toString());
            item->setBackground(QBrush(color));
            m_table->setItem(row, col, item);
        }
    }
    m_countLabel->setText(QString("%1 objects").arg(filtered.size()));
}

File: gui/widgets/object_table_page.py
"""
Page 4: Object Dictionary Table (dev doc §12.4)

Shows all objects in a filterable table with color coding:
  green=auto, yellow=callback, red=error, gray=skipped
"""
from PyQt5.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QTableWidget, QTableWidgetItem,
    QComboBox, QLabel, QHeaderView, QGroupBox,
)
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QColor, QBrush


# Color scheme per dev doc §12.4
_COLORS = {
    "auto":     QColor(200, 255, 200),   # green
    "callback": QColor(255, 255, 180),   # yellow
    "error":    QColor(255, 200, 200),   # red
    "skipped":  QColor(220, 220, 220),   # gray
    "normal":   QColor(255, 255, 255),   # white
}

_TABLE_HEADERS = [
    "Index", "Name", "SubCount", "Category",
    "PDO Map", "SM Obj", "Risk", "CoeRead", "CoeWrite",
]


class ObjectTablePage(QWidget):
    """Object dictionary table with filtering."""

    def __init__(self, parent=None):
        super().__init__(parent)
        self._objects = []  # cached object list
        self._init_ui()

    def _init_ui(self):
        layout = QVBoxLayout(self)

        # ── Filter bar ─────────────────────────
        filter_row = QHBoxLayout()
        filter_row.addWidget(QLabel("Filter:"))
        self._filter_combo = QComboBox()
        self._filter_combo.addItems([
            "All", "Errors only", "Callback only",
            "PDO mapping only", "Fxxx only", "Skipped",
        ])
        self._filter_combo.currentIndexChanged.connect(self._apply_filter)
        filter_row.addWidget(self._filter_combo)
        filter_row.addStretch()
        self._count_label = QLabel("0 objects")
        filter_row.addWidget(self._count_label)
        layout.addLayout(filter_row)

        # ── Table ──────────────────────────────
        self._table = QTableWidget()
        self._table.setColumnCount(len(_TABLE_HEADERS))
        self._table.setHorizontalHeaderLabels(_TABLE_HEADERS)
        self._table.setEditTriggers(QTableWidget.NoEditTriggers)
        self._table.setSelectionBehavior(QTableWidget.SelectRows)
        self._table.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        layout.addWidget(self._table)

    # ── Public API ─────────────────────────────

    def set_objects(self, objects: list[dict]):
        """Load objects into the table. Each dict has keys matching _TABLE_HEADERS."""
        self._objects = objects
        self._apply_filter()

    def _apply_filter(self):
        """Filter and refresh the table."""
        filter_idx = self._filter_combo.currentIndex()
        filtered = self._objects

        if filter_idx == 1:   # Errors only
            filtered = [o for o in filtered if o.get("risk")]
        elif filter_idx == 2: # Callback only
            filtered = [o for o in filtered if o.get("CoeRead") or o.get("CoeWrite")]
        elif filter_idx == 3: # PDO mapping
            filtered = [o for o in filtered if o.get("pdo_map")]
        elif filter_idx == 4: # Fxxx
            filtered = [o for o in filtered if o.get("index_int", 0) >= 0xF000]
        elif filter_idx == 5: # Skipped
            filtered = [o for o in filtered if o.get("skipped")]

        self._populate_table(filtered)

    def _populate_table(self, objects: list[dict]):
        """Fill the QTableWidget with object data."""
        self._table.setRowCount(len(objects))
        for row, obj in enumerate(objects):
            color = _determine_color(obj)
            brush = QBrush(color)

            for col, key in enumerate(_TABLE_HEADERS):
                val = str(obj.get(key, ""))
                item = QTableWidgetItem(val)
                item.setBackground(brush)
                self._table.setItem(row, col, item)

        self._count_label.setText(f"{len(objects)} objects")


def _determine_color(obj: dict) -> QColor:
    """Determine row color based on object state."""
    if obj.get("skipped"):
        return _COLORS["skipped"]
    if obj.get("risk"):
        return _COLORS["error"]
    if obj.get("CoeRead") or obj.get("CoeWrite"):
        return _COLORS["callback"]
    if obj.get("category") == "comm":
        return _COLORS["auto"]
    return _COLORS["normal"]
